va/va.go

```diff
@@ -254,96 +256,76 @@ func (va VAImpl) validateTLSSNI02(task *vaTask) *core.ValidationRecord {
-	zb := hex.EncodeToString(hb[:])
-	sanBName := fmt.Sprintf("%s.%s.%s.%s", zb[:32], zb[32:], tlsSNIKaID, tlsSNISuffix)
-
-	// Perform the validation
-	result.Error = va.validateTLSSNI02WithNames(hostPort, sanAName, sanBName)
-	return result
-}
-
-func (va VAImpl) validateTLSSNI02WithNames(hostPort string, sanAName, sanBName string) *acme.ProblemDetails {
-	certs, problem := va.fetchCerts(hostPort, sanAName)
+	cs, problem := va.fetchConnectionState(hostPort, &tls.Config{
+		ServerName:         task.Identifier,
+		NextProtos:         []string{acme.ACMETLS1Protocol},
+		InsecureSkipVerify: true,
+	})
 	if problem != nil {
-		return problem
+		result.Error = problem
+		return result
 	}
 
-	leafCert := certs[0]
-	if len(leafCert.DNSNames) != 2 {
-		names := certNames(leafCert)
-		msg := fmt.Sprintf(
-			"%s challenge certificate doesn't include exactly 2 DNSName entries. "+
-				"Received %d certificate(s), first certificate had names %q",
-			acme.ChallengeTLSSNI02, len(certs), names)
-		return acme.MalformedProblem(msg)
+	if !cs.NegotiatedProtocolIsMutual || cs.NegotiatedProtocol != acme.ACMETLS1Protocol {
+		result.Error = acme.UnauthorizedProblem(fmt.Sprintf(
+			"Cannot connect using %s for %s challenge",
+			acme.ACMETLS1Protocol,
+			acme.ChallengeTLSALPN01,
+		))
+		return result
 	}
 
-	var validSanAName, validSanBName bool
-	for _, name := range leafCert.DNSNames {
-		if subtle.ConstantTimeCompare([]byte(name), []byte(sanAName)) == 1 {
-			validSanAName = true
-		}
-
-		if subtle.ConstantTimeCompare([]byte(name), []byte(sanBName)) == 1 {
-			validSanBName = true
-		}
+	certs := cs.PeerCertificates
+	if len(certs) == 0 {
+		result.Error = acme.UnauthorizedProblem(fmt.Sprintf("No certs presented for %s challenge", acme.ChallengeTLSALPN01))
+		return result
 	}
+	leafCert := certs[0]
 
-	if !validSanAName || !validSanBName {
+	// Verify SNI
+	if len(leafCert.DNSNames) != 1 || !strings.EqualFold(leafCert.DNSNames[0], task.Identifier) {
 		names := certNames(leafCert)
-		msg := fmt.Sprintf(
+		errText := fmt.Sprintf(
 			"Incorrect validation certificate for %s challenge. "+
 				"Requested %s from %s. Received %d certificate(s), "+
 				"first certificate had names %q",
-			acme.ChallengeTLSSNI02, sanAName, hostPort,
-			len(certs), names)
-		return acme.UnauthorizedProblem(msg)
+			acme.ChallengeTLSALPN01, task.Identifier, hostPort, len(certs), names)
+		result.Error = acme.UnauthorizedProblem(errText)
+		return result
 	}
 
-	return nil
+	// Verify key authorization in acmeValidation extension
+	expectedKeyAuthorization := task.Challenge.ExpectedKeyAuthorization(task.Account.Key)
+	h := sha256.Sum256([]byte(expectedKeyAuthorization))
+	for _, ext := range leafCert.Extensions {
+		if IdPeAcmeIdentifierV1.Equal(ext.Id) && ext.Critical {
+			if subtle.ConstantTimeCompare(h[:], ext.Value) == 1 {
+				return result
+			}
+			result.Error = acme.UnauthorizedProblem("Extension acmeValidationV1 value incorrect.")
```

Comment:
I think it might be clearer if this error string matched the format of the `errText` above, e.g: "Incorrect validation certificate for %s challenge. Missing expected acmeValidation1 extension value"

Reply:
Agreed! Good catch :-)